### PR TITLE
[#56] Refactor: NextJs AppDir

### PR DESCRIPTION
**Pull request by Drish-xD**
Issues Fixed: #56 #57 #45 #15 #18 #19 #20 #25 

# Tasks: 
- [x] Convert to NextJs AppDir 
- [x] Add SchadCn UI lib
- [x] Fix Types used in app
- [x] Data Table Improvements
- [x] Form Stepper Improvements
- [x] Form Builder for easy form creation
- [x] Server Action Addition

### CODE REVIEW

File: src/Constants/Options.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+export const TestTypeOptions: Option[] = [
+  { value: 'assessment', label: 'Assessment' },
+  { value: 'homework', label: 'Homework' },
+  { value: 'omr-assessment', label: 'Omr-Assessment' },
+];
+
+export const SessionTypeOptions: Option[] = [
+  { value: 'sign-in', label: 'SignIn' },
+  { value: 'sign-up', label: 'SignUp' },
+  { value: 'sign-in with forgot id', label: 'SignIn with forgot id' },
```
Comment (suryabulusu) [inline src/Constants/Options.ts:12]: "Forgot ID" capitalize

File: src/Constants/Options.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+];
+
+export const OptionalLimitOptions: Option[] = [
+  { value: 'NA', label: 'NA' },
+  { value: 'neet', label: 'NEET' },
+  { value: 'jee', label: 'JEE' },
+  { value: 'cuet', label: 'CUET' },
+];
+
+export const GroupOptions: Option[] = [
```
Comment (suryabulusu) [inline src/Constants/Options.ts:51]: almost entirely replaced later on

File: src/Constants/Options.ts [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  { value: 'HimachalStudents', label: 'Himachal Students' },
+  { value: 'DelhiStudents', label: 'Delhi Students' },
+  { value: 'EnableStudents', label: 'Enable Students' },
+  { value: 'AFTeachers', label: 'AF Teachers' },
+  { value: 'Candidates', label: 'Candidates' },
+  { value: 'GujaratStudents', label: 'Gujarat Students' },
+  { value: 'AFTesting', label: 'AF Testing' },
+];
+
+export const BatchOptions: Option[] = [
```
Comment (suryabulusu) [inline src/Constants/Options.ts:62]: almost entirely replaced later on

File: src/Constants/Options.ts [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+];
+
+export const GradeOptions = Grades.map((g) => ({ value: g, label: g }));
+
+export const AuthOptions: Option[] = Object.values(AuthType).map((item) => ({
+  value: item,
+  label: item,
+}));
+
+export const ActiveDaysOptions: Option[] = [
```
Comment (suryabulusu) [inline src/Constants/Options.ts:218]: i feel we can provide few presets here later on. 

all selected ... MWF selected. .. etc

File: src/Constants/index.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+export const PAGE_SIZE_OPTIONS = [10, 20, 30, 40, 50];
+
+export const KeysToDeleteBeforeDuplicate: DeepKeys<Session>[] = [
+  'end_time',
+  'start_time',
+  'name',
+  'id',
+  'portal_link',
+  'signup_form_id',
+  'owner_id',
```
Comment (suryabulusu) [inline src/Constants/index.ts:16]: why is owner id part of code..

File: src/Constants/index.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+export const DATA_PER_PAGE = 10;
+export const PAGE_SIZE_OPTIONS = [10, 20, 30, 40, 50];
+
+export const KeysToDeleteBeforeDuplicate: DeepKeys<Session>[] = [
+  'end_time',
+  'start_time',
+  'name',
+  'id',
+  'portal_link',
+  'signup_form_id',
```
Comment (suryabulusu) [inline src/Constants/index.ts:15]: i think we can retain this

Comment (suryabulusu) [reply src/Constants/index.ts:15]: provided signup_form is true

Comment (suryabulusu) [reply src/Constants/index.ts:15]: but for now its fine

File: src/Constants/index.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  'owner_id',
+  'session_id',
+  'platform_link',
```
Comment (suryabulusu) [inline src/Constants/index.ts:18]: note that we cannot retain this for liveclass flow.. 
same goes for platform_id .

File: src/Constants/index.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  'platform_id',
+  'created_by_id',
+  'inserted_at',
+  'updated_at',
+  'popup_form_id',
```
Comment (suryabulusu) [inline src/Constants/index.ts:23]: can avoid retaining this in next PR

File: src/app/(home)/Table/Columns.tsx [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+    title: 'Basic Details',
+    data: [
+      { label: 'Platform', value: data.platform },
+      { label: 'Program', value: data.meta_data?.group },
+      { label: 'Batch', value: data.meta_data?.batch },
+      { label: 'Grade', value: data.meta_data?.grade },
+      { label: 'Session Type', value: data.type },
+      { label: 'Auth Type', value: data.auth_type },
+      { label: 'Activate Sign-Up', value: data.is_active ? 'Yes' : 'No' },
+      { label: 'Is pop up form allowed', value: data.popup_form ? 'Yes' : 'No' },
```
Comment (suryabulusu) [inline src/app/(home)/Table/Columns.tsx:172]: Is Pop Up Form Allowed?

File: src/app/(home)/Table/Columns.tsx [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+      {
+        label: 'Number fields in pop up form',
```
Comment (suryabulusu) [inline src/app/(home)/Table/Columns.tsx:174]: Number of fields in Popup Form

File: src/app/(home)/Table/Columns.tsx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+        value: data.meta_data?.number_of_fields_in_popup_form,
+      },
```
Comment (suryabulusu) [inline src/app/(home)/Table/Columns.tsx:176]: Is Redirection allowed?

File: src/app/(home)/Table/Columns.tsx [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+      { label: 'Is redirection allowed', value: data.redirection ? 'Yes' : 'No' },
+      { label: 'Is Id generation allowed', value: data.id_generation ? 'Yes' : 'No' },
```
Comment (suryabulusu) [inline src/app/(home)/Table/Columns.tsx:178]: Is ID Generation allowed?

File: src/app/(home)/Table/Columns.tsx [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+      { label: 'Signup form name', value: data.meta_data?.signup_form_name ?? 'N/A' },
```
Comment (suryabulusu) [inline src/app/(home)/Table/Columns.tsx:179]: Signup Form Name

Comment (suryabulusu) [reply src/app/(home)/Table/Columns.tsx:179]: note that this is displayed only if signup_form is true etc .. if signup_form is false then no need to show.

infact if session type is "sign-in" or "sign-in with forgot id" .. signup_form should be false and signup_form_name option shoudln't be displayed at all. 

keep in mind for later if not this PR. make an issue in github repo

File: src/app/(home)/Table/Columns.tsx [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+            { label: 'Test Purpose', value: data.meta_data?.test_purpose },
+            { label: 'Test Type', value: data.meta_data?.test_type },
+            {
+              label: 'CMS Link',
+              value: data.meta_data?.cms_test_id ?? 'N/A',
+              isLink: !!data.meta_data?.cms_test_id,
+            },
+            { label: 'Marking Scheme', value: data.meta_data?.marking_scheme },
+            { label: 'Optional Limits', value: data.meta_data?.optional_limits },
+            { label: 'Show Answers', value: data.meta_data?.show_answers ? 'Yes' : 'No' },
```
Comment (suryabulusu) [inline src/app/(home)/Table/Columns.tsx:201]: why is this not getting deleted in duplicate columns

File: src/app/session/[type]/Steps/Basic/index.tsx [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        type: 'select',
+        options: SessionTypeOptions,
+        placeholder: 'Select a session type',
+        label: 'Session type',
+      },
+      authType: {
+        type: 'select',
+        options: AuthOptions,
+        label: 'Auth type',
+        placeholder: 'Select a auth type',
```
Comment (suryabulusu) [inline src/app/session/[type]/Steps/Basic/index.tsx:63]: an

File: src/app/session/[type]/Steps/Basic/index.tsx [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+      },
+      activateSignUp: {
+        type: 'switch',
+        label: 'Activate sign up',
```
Comment (suryabulusu) [inline src/app/session/[type]/Steps/Basic/index.tsx:67]: Activate Sign Up .. just check caps throughout

File: src/app/session/[type]/Steps/Basic/index.tsx [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+        helperText: 'Do you want to display sign up form?',
+      },
+      isPopupForm: {
+        type: 'switch',
+        label: 'Is pop up form allowed',
```
Comment (suryabulusu) [inline src/app/session/[type]/Steps/Basic/index.tsx:72]: Popup

File: src/app/session/page.tsx [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+export default async function ViewSession({ searchParams: { id } }: Props) {
+  if (!id || typeof parseInt(id) !== 'number') {
+    notFound();
+  }
+
+  const sessionData = await getASession(Number(id) || null);
+
+  return (
+    <main className="container text-center">
+      <h2>TO BE IMPLEMENTED {(sessionData as Session)?.name}</h2>
```
Comment (suryabulusu) [inline src/app/session/page.tsx:18]: whats this for

File: src/components/ui/checkbox.tsx [hunk 19]
```diff
@@ -0,0 +1,76 @@
+'use client';
```
Comment (suryabulusu) [inline src/components/ui/checkbox.tsx:1]: for these ui tsx files.. follow the same naming convention na.. capitalize the file name? Card.tsx, Checkbox.tsx etc..

File: src/lib/utils.ts [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+        newDefaultData = {
+          repeat_schedule: {
+            type: 'weekly',
+            params: [1, 2, 3, 4, 5, 6, 7],
+          },
+          type: 'sign-in',
+          auth_type: AuthType.ID,
+          id_generation: false,
+          redirection: false,
+          signup_form: true,
```
Comment (suryabulusu) [inline src/lib/utils.ts:51]: why are signup_form_name / signup_form_id not here

Comment (suryabulusu) [reply src/lib/utils.ts:51]: check for other cases too

File: src/services/services.ts [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+      session_id: '',
+      meta_data: {
+        ...formData.meta_data,
+        report_link: '',
+        shortened_link: '',
+        has_synced_to_bq: false,
+        infinite_session: false,
+        date_created: new Date(),
+      },
+      purpose: {
```
Comment (suryabulusu) [inline src/services/services.ts:75]: this will change for liveclass

File: src/services/services.ts [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+ * - id : Id of the created session
+ */
+export async function createSession(formData: Session) {
+  try {
+    const payload: Session = {
+      ...formData,
+      session_id: '',
+      meta_data: {
+        ...formData.meta_data,
+        report_link: '',
```
Comment (suryabulusu) [inline src/services/services.ts:69]: wont be there for non quiz platform.. next pr

**Review by suryabulusu — COMMENTED** [review]
left minor comments.. will try to launch on amplify now

File: src/Constants/Options.ts (hunk 1 above)
Comment (Drish-xD) [reply src/Constants/Options.ts:12]: Should I Capitalize in label only or the value also?

File: src/Constants/Options.ts [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+  { value: 'weekly_test', label: 'Weekly Test' },
+  { value: 'monthly_test', label: 'Monthly Test' },
+  { value: 'reshuffling_test', label: 'Reshuffling Test' },
+  { value: 'selection_test', label: 'Selection Test' },
+  { value: 'one_time_test', label: 'One Time Test' },
+];
+
+export const TestPlatformOptions = [
+  { value: 'quiz', label: 'Quiz' },
+  { value: 'live', label: 'Live Class' },
```
[the remark this answers is not in the document]
Comment (Drish-xD) [reply src/Constants/Options.ts:36]: Done in the new PR #55
.

File: src/Constants/Options.ts (hunk 2 above)
Comment (Drish-xD) [reply src/Constants/Options.ts:51]: Done in new PR #55, fetched from db.

File: src/Constants/Options.ts (hunk 3 above)
Comment (Drish-xD) [reply src/Constants/Options.ts:62]: Done in new PR #55, fetched from db.

**Review by suryabulusu — APPROVED** [review]
looks good

**Pull request merged**